sql/src/main/java/io/crate/execution/dml/upsert/ShardInsertRequest.java

```diff
@@ -0,0 +1,371 @@
+        }
+        ShardInsertRequest items = (ShardInsertRequest) o;
+        return Objects.equals(sessionSettings, items.sessionSettings) &&
+               Arrays.equals(insertColumns, items.insertColumns) &&
+               Objects.equals(values, items.values);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(super.hashCode(), sessionSettings, values);
+        result = 31 * result + Arrays.hashCode(insertColumns);
+        return result;
+    }
+
+    // Values are only used for internal storage and serialization
+    private enum Values {
+        DUPLICATE_KEY_UPDATE_OR_FAIL,
+        DUPLICATE_KEY_OVERWRITE,
+        DUPLICATE_KEY_IGNORE,
+        CONTINUE_ON_ERROR,
+        VALIDATE_CONSTRAINTS;
+
+        static EnumSet<Values> toEnumSet(boolean continueOnError, boolean validateConstraints, DuplicateKeyAction action) {
+            ArrayList<Values> values = new ArrayList<>(3);
+            if (continueOnError) {
+                values.add(Values.CONTINUE_ON_ERROR);
+            }
+            if (validateConstraints) {
+                values.add(Values.VALIDATE_CONSTRAINTS);
+            }
+            switch (action) {
+                case IGNORE:
+                    values.add(Values.DUPLICATE_KEY_IGNORE);
+                    break;
+                case OVERWRITE:
+                    values.add(Values.DUPLICATE_KEY_OVERWRITE);
+                    break;
+                case UPDATE_OR_FAIL:
+                    values.add(Values.DUPLICATE_KEY_UPDATE_OR_FAIL);
+                    break;
+                default:
+                    throw new IllegalArgumentException("DuplicateKeyAction not supported for serialization: " + action.name());
+            }
+            return EnumSet.copyOf(values);
+        }
+
+        static DuplicateKeyAction duplicationAction(EnumSet<Values> values) {
+            if (values.contains(Values.DUPLICATE_KEY_UPDATE_OR_FAIL)) {
+                return DuplicateKeyAction.UPDATE_OR_FAIL;
+            }
+            if (values.contains(Values.DUPLICATE_KEY_OVERWRITE)) {
+                return DuplicateKeyAction.OVERWRITE;
+            }
+            if (values.contains(Values.DUPLICATE_KEY_IGNORE)) {
+                return DuplicateKeyAction.IGNORE;
+            }
+            throw new IllegalArgumentException("DuplicateKeyAction not found");
+        }
+
+        static boolean continueOnError(EnumSet<Values> values) {
+            return values.contains(Values.CONTINUE_ON_ERROR);
+        }
+
+        static boolean validateConstraints(EnumSet<Values> values) {
+            return values.contains(Values.VALIDATE_CONSTRAINTS);
+        }
+    }
+
+    /**
+     * A single insert item.
+     */
+    public static final class Item extends ShardWriteRequest.Item {
+
+        @Nullable
+        protected BytesReference source;
+
+        /**
+         * List of objects used on insert
+         */
+        @Nullable
+        private Object[] insertValues;
```

Comment:
Do we need both `source` and `insertValues`?

I assume `insertValues` is for the initial primary operation, and `source` for the replica request?

Would it make sense to split the `Item` into `PrimaryItem` and `ReplicaItem` ?